### PR TITLE
Add pipeline source: HTTP

### PR DESCRIPTION
**Pull request by julienduchesne**
Allows fetching a Jenkinsfile from a given HTTP URL
Basic auth credentials can be set
Usecases:
- SCM Host APIs provide full files (faster than checkout)
- Amazon S3 (or other cloud file storage providers)

image

### CODE REVIEW

**Comment by julienduchesne** [comment]
Our personal use case is that we pull the same repo from bitbucket hundreds of times on the same Jenkins instance. I think we are starting to hit a limit with all those concurrent checkouts. We want to switch to using the Bitbucket API instead

**Comment by julienduchesne** [comment]
@jenkinsci/code-reviewers. What do you think?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2019 CloudBees, Inc.
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java]: Needs to be changed if you do not work for CloudBees

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+        String expandedScriptUrl = build.getEnvironment(listener).expand(scriptUrl);
+        URL url = new URL(expandedScriptUrl);
+        listener.getLogger().println("Fetching pipeline from " + expandedScriptUrl);
+        int count = 0;
+        int maxTries = retryCount + 1;
+
+        while (true) {
+            try {
+                HttpURLConnection con = (HttpURLConnection) url.openConnection();
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java]: https://github.com/jenkinsci/apache-httpcomponents-client-4-api-plugin/blob/master/src/main/java/io/jenkins/plugins/httpclient/RobustHTTPClient.java could be used for better stability

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java (hunk 1 above)
Comment (julienduchesne) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java]: I just write my name or nothing at all?

Comment (oleg-nenashev) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java]: Your name, your company name, or whatever "Plugin contributors" if you prefer.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinitionTest.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+public class CpsHttpFlowDefinitionTest {
+    @Rule public JenkinsRule r = new JenkinsRule();
+
+    @Test public void testRunJenkinsHomePageAsPipeline() throws Exception {
+        WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+        CpsHttpFlowDefinition def = new CpsHttpFlowDefinition(r.jenkins.getRootUrl(), "", 3);
+        p.setDefinition(def);
+        WorkflowRun b = r.assertBuildStatus(Result.FAILURE, p.scheduleBuild2(0));
+        r.assertLogContains("unexpected token", b);
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinitionTest.java]: Should be at least one _positive_ test! You can use the `userContent` system for example.

File: src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition/help-scriptUrl.html [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<div>
+    Full URL to your script
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition/help-scriptUrl.html]:
```suggestion
    Full URL to your script.
```

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+@PersistIn(JOB) public class CpsHttpFlowDefinition extends FlowDefinition {
+
+    private final String scriptUrl;
+    private final String credentialsId;
+    private final int retryCount;
+
+    @DataBoundConstructor public CpsHttpFlowDefinition(String scriptUrl, String credentialsId, int retryCount) {
+        this.scriptUrl = scriptUrl.trim();
+        this.credentialsId = credentialsId;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java]:
```suggestion
        this.credentialsId = Util.fixEmpty(credentialsId);
```

and later you can simply check for null.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+import static org.jenkinsci.plugins.workflow.cps.persistence.PersistenceContext.JOB;
+
+@PersistIn(JOB) public class CpsHttpFlowDefinition extends FlowDefinition {
+
+    private final String scriptUrl;
+    private final String credentialsId;
+    private final int retryCount;
+
+    @DataBoundConstructor public CpsHttpFlowDefinition(String scriptUrl, String credentialsId, int retryCount) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java]: Generally speaking, optional parameters should be moved to a `@DataBoundSetter`.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java (hunk 2 above)
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java]: And easier authentication probably.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        int count = 0;
+        int maxTries = retryCount + 1;
+
+        while (true) {
+            try {
+                HttpURLConnection con = (HttpURLConnection) url.openConnection();
+                con.setRequestMethod("GET");
+
+                if (!StringUtils.isBlank(credentialsId)) {
+                    UsernamePasswordCredentials credentials = (UsernamePasswordCredentials) CredentialsMatchers.firstOrNull(CredentialsProvider.lookupCredentials(UsernamePasswordCredentials.class, Jenkins.getInstance(), ACL.SYSTEM, Collections.emptyList()), CredentialsMatchers.withId(credentialsId));
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java]: IIRC there is some additional API that should be called to record that this build is using these credentials.

**Comment by jglick** [comment]
> we are starting to hit a limit with all those concurrent checkouts. We want to switch to using the Bitbucket API instead

So that would be better handled by just selecting the `CpsScmFileSystem.lightweight` option, assuming the BitBucket SCM plugin implements `SCMFileSystem` (if not, fix it to do so).

**Comment by jglick** [comment]
or [this way](https://github.com/AmadeusITGroup/workflow-cps-global-lib-http-plugin#directly-in-the-jenkinsfile)

**Comment by julienduchesne** [comment]
> Is this a sufficiently common need that it deserves to be in `workflow-cps`? Most people would use the SCM script source. Could be done as a separate plugin with an independent maintainer I think.

I think it's generic enough to be there but that is my opinion. If you won't accept, it will be a plugin of course

> > we are starting to hit a limit with all those concurrent checkouts. We want to switch to using the Bitbucket API instead
> 
> So that would be better handled by just selecting the `CpsScmFileSystem.lightweight` option, assuming the BitBucket SCM plugin implements `SCMFileSystem` (if not, fix it to do so).

We are using interpolation in our branch name which the lightweight doesn't support. Anyhow, HTTP is much much faster.


> or [this way](https://github.com/AmadeusITGroup/workflow-cps-global-lib-http-plugin#directly-in-the-jenkinsfile)

Libraries cannot be used as a pipeline entrypoint. You need an inital Jenkinsfile to pull a library

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java (hunk 2 above)
Comment (julienduchesne) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java]: I haven't found any usage of this anywhere. Do you have any examples?

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsHttpFlowDefinition.java]: [query](https://github.com/search?q=user%3Ajenkinsci+RobustHTTPClient&type=Code)

**Comment by julienduchesne** [comment]
Alright. The changes have been made. So, what do you guys think about whether or not this belongs in this repo?

**Comment by daniel-beck** [comment]
Can easily be done as a separate plugin, that's how I did it with https://plugins.jenkins.io/inline-pipeline

**Comment by julienduchesne** [comment]
Of course, I'm just trying to understand at what point something should be integrated or be made as a separate plugin. Does it make sense to create a new plugin to host a single 100-line code file when that code is not any less generic than the SCM implementation?

Also, any pointers on how to add a plugin to the `jenkinsci` org in github? If I split it, I'd like it to be an "official" plugin

**Comment by julienduchesne** [comment]
Alright. I migrated the plugin here: https://github.com/coveo/workflow-cps-http-plugin

**Comment by dwnusbaum** [comment]
> Also, any pointers on how to add a plugin to the jenkinsci org in github?

@julienduchesne see https://jenkins.io/doc/developer/publishing/requesting-hosting/. Note also that someone recently created https://github.com/jenkinsci/workflow-cps-global-lib-http-plugin, maybe it makes sense for these features to live in the same plugin.

**Pull request closed without merging**